pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,517 @@
+	case []apiv2.PodGroupMemberSpec:
+		for _, pp := range p {
+			count += s.countLeafCellNums(pp)
+		}
+	case *apiv2.PodGroupSpec:
+		count += s.countLeafCellNums(p.Pods) + s.countLeafCellNums(p.ChildGroups)
+	case []*apiv2.PodGroupSpec:
+		for _, pp := range p {
+			count += s.countLeafCellNums(pp)
+		}
+	}
+	return count
+}
+
+func (s *skuScheduler) findCellsForPodGroup(
+	podGroup *apiv2.PodGroupSpec,
+	priority CellPriority,
+	within *skuCell,
+	allocated *PodGroupPlacement) (
+	placement PodGroupPlacement,
+	failedReason string) {
+
+	placement, failedReason = PodGroupPlacement{}, "no matched cells in vc"
+
+	cv := skuClusterView{nil}
+	if level, ok := s.cellLevels[podGroup.WithinOneCell]; ok {
+		cv = s.createSkuClusterView(within, level, priority)
+	} else if within != nil {
+		cv = s.createSkuClusterView(within, within.cell.GetLevel(), priority)
+	}
+
+	for _, withinCell := range cv {
+		if len(podGroup.Pods) > 0 && withinCell != nil && !withinCell.healthy {
+			return PodGroupPlacement{}, fmt.Sprintf(
+				"have to use at least one bad cell %v", withinCell.address)
+		}
+		placement.podsPlacement, failedReason = s.findCellsForPods(podGroup.Pods, priority, withinCell, allocated)
+		if failedReason == "" {
+			for _, childGroup := range podGroup.ChildGroups {
+				childPodsPlacement, childFailedReason := s.findCellsForPodGroup(childGroup, priority, withinCell, &placement)
+				if childFailedReason != "" {
+					placement.childGroupsPlacement, failedReason = nil, childFailedReason
+					break
+				}
+				if placement.childGroupsPlacement == nil {
+					placement.childGroupsPlacement = []*PodGroupPlacement{}
+				}
+				placement.childGroupsPlacement = append(placement.childGroupsPlacement, &childPodsPlacement)
+			}
+			if failedReason == "" {
+				break
+			}
+		}
+	}
+	return placement, failedReason
+}
+
+func (s *skuScheduler) findCellsForPods(
+	pods []apiv2.PodGroupMemberSpec,
+	priority CellPriority,
+	within *skuCell,
+	allocated *PodGroupPlacement) (
+	placement []CellList,
+	failedReason string) {
+
+	placement, failedReason = []CellList{}, ""
+
+	allocatedCells := CellList{}
+	for iter := allocated.Iterator(); iter.HasNext(); {
+		for _, c := range *iter.Next() {
+			allocatedCells = append(allocatedCells, c)
+		}
+	}
+
+	cv := skuClusterView{within}
+	nodeLevel := s.getNodeLevel()
+	if within == nil || within.cell.GetLevel() > nodeLevel {
+		cv = s.createSkuClusterView(within, nodeLevel, priority)
+	}
+
+	withinCellIndex, podIndex := 0, 0
+	for podIndex < len(pods) {
+		if withinCellIndex >= len(cv) {
+			return nil, "insufficient capacity"
+		}
+		withinCell := cv[withinCellIndex]
+		if !withinCell.healthy {
+			return nil, fmt.Sprintf(
+				"have to use at least one bad cell %v", withinCell.address)
+		}
+		podPlacement := s.findCellsForSinglePod(pods[podIndex], priority, withinCell, allocatedCells)
+		if podPlacement == nil {
+			withinCellIndex++
```

Comment:
>withinCellIndex [](http://example.com/codeflow?start=3&length=15)

Do we need to reset withinCellIndex = 0 before schedule each pod?

Reply:
since we have changed `podGroup.Pods` to `podGroup.Pod`, all `pods` here with different `podIndex` are the same, no need to reset `withinCellIndex` anymore